File: TrappingRainWater.java
/*
Runtime: 1 ms, faster than 99.76% of Java online submissions for Trapping Rain Water.
Memory Usage: 42.9 MB, less than 93.35% of Java online submissions for Trapping Rain Water.
*/
// Best Solution TC: O(N) SC: O(1)
class Solution {
    public int trap(int[] height) {
        
        // Declaring 2 Pointers At Start And End
        int left = 0;
        int right = height.length - 1;
        
        // For Storing Result
        int result = 0;
        
        // For Storing Maximum From Left Side And Right Side In The Array While Traversing Using The Two Pointers 
        int leftMax = 0;
        int rightMax = 0;
        
        // Iterating Using 2 Pointers
        while(left < right){
            
            // If Height At Left Pointer Is Smaller Than Height At Right Pointer
            if(height[left] < height[right]){
                
                // We Check If Height At Left Is Larger Than leftMax
                // If It Is, Then We Store The Value Of Height At leftMax
                // Else, We Add The leftMax - Height Value To Amount Of Water
                // We Do The Same For The Right Pointer
                if(height[left] >= leftMax){
                    leftMax = height[left];
                }
                else{
                    result += (leftMax - height[left]);
                }
                
                left++;
                
            }
            else{
                
                if(height[right] >= rightMax){
                    rightMax = height[right];
                }
                else{
                    result += (rightMax - height[right]);
                }
                
                right--;
                
            }
            
        }
        
        // Returning Result
        return result;
        
    }
}



/*
Runtime: 1 ms, faster than 99.76% of Java online submissions for Trapping Rain Water.
Memory Usage: 48.8 MB, less than 52.21% of Java online submissions for Trapping Rain Water.
*/
// Optimized Solution TC: O(n) SC: O(n)
class Solution {
    public int trap(int[] height) {
        
        int heightLength = height.length;
        int[] leftMaxSize = new int[heightLength];
        int[] rightMaxSize = new int[heightLength];
        
        // Storing The Total Amount Of Water 
        int amountOfWaterStored = 0;
        
        /*
        Approach:
          1. We Calculate The Max Height From Left And Right, Then Taking Their Minimum And Subtracting It
             With The Current Height Will Give Us The Amount Of Water Stored Above The Current Element.
          2. Then We Keep Adding The Amount Of Water To The Total Amount Of Water And Finally We Return The 
             Answer.
        */
        int currentMax = 0;
        
        for(int i=0; i<heightLength; i++){
            if(currentMax < height[i]){
                currentMax = height[i];
            }
            leftMaxSize[i] = currentMax;
        }
        
        currentMax = 0;
        
        for(int i=heightLength-1; i>=0; i--){
            if(currentMax < height[i]){
                currentMax = height[i];
            }
            rightMaxSize[i] = currentMax;
        }
        
        for(int i=0; i<heightLength; i++){
            amountOfWaterStored += Math.min(leftMaxSize[i], rightMaxSize[i]) - height[i];
        }
        
        // Returning The Total Amount Of Water Stored.
        return amountOfWaterStored;
        
    }
}

File: TrappingRainWater.js
/**
 * @param {number[]} 
 * @return {number} 
 */
// TC: O(N) SC: O(1)
var trap = function(height) {
    // Initialize two pointers, one at the beginning and one at the end.
    let left = 0;
    let right = height.length - 1;

    // Initialize variables to keep track of the maximum height encountered from each side.
    let leftMax = 0;
    let rightMax = 0;

    // Initialize the total trapped water to zero.
    let totalWater = 0;

    // Iterate while the left pointer is less than the right pointer.
    while (left < right) {
        // If the current left bar is smaller than the current right bar,
        // it means the left bar determines the trapping capacity.
        if (height[left] < height[right]) {
            // If the current left bar is greater than or equal to the current leftMax,
            // update leftMax.
            if (height[left] >= leftMax) {
                leftMax = height[left];
            } else {
                // Otherwise, add the trapped water to the total.
                // The water trapped is the difference between the current leftMax and the current left bar's height.
                totalWater += leftMax - height[left];
            }
            // Move the left pointer one step to the right.
            left++;
        } else {
            // If the current right bar is smaller than or equal to the current left bar,
            // it means the right bar determines the trapping capacity.
            // Apply the same logic as above but for the right side.
            if (height[right] >= rightMax) {
                rightMax = height[right];
            } else {
                totalWater += rightMax - height[right];
            }
            // Move the right pointer one step to the left.
            right--;
        }
    }

    // Return the total trapped water.
    return totalWater;
};


//TC: O(N) SC: O(N)
var trap = function(height) {
    const left = [];  // Array to store maximum heights encountered from the left
    const right = []; // Array to store maximum heights encountered from the right

    // Precompute maximum heights from the left:
    for (let i = 0; i < height.length; i++) {
        if (i === 0) {
            left[i] = height[i]; // Base case: first element is its own left max
        } else {
            left[i] = Math.max(height[i], left[i - 1]); // Store the maximum between the current height and the previous left max
        }
    }

    // Precompute maximum heights from the right:
    for (let i = height.length - 1; i >= 0; i--) {
        if (i === height.length - 1) {
            right[i] = height[i]; // Base case: last element is its own right max
        } else {
            right[i] = Math.max(right[i + 1], height[i]); // Store the maximum between the current height and the previous right max
        }
    }

    // Calculate trapped water:
    let result = 0;
    for (let i = 0; i < height.length; i++) {
        // Water is limited by the shorter of the left and right max heights
        const waterLevel = Math.min(left[i], right[i]);
        result += waterLevel - height[i]; 
    }

    return result;
};
